File: 3714-longest-balanced-substring-ii/3714-longest-balanced-substring-ii.go
func longestBalanced(s string) int {
    res := []int{
        solve1(s, 'a'),
        solve1(s, 'b'),
        solve1(s, 'c'),
        solve2(s, 'a', 'b'),
        solve2(s, 'a', 'c'),
        solve2(s, 'b', 'c'),
        solve3(s),
    }
    
    ans := 0
    for _, r := range res {
        ans = max(ans, r)
    }
    return ans
}

// hash map, prefix sum, string
// ref: tdkkdt
// time: O(n)
// space: O(n)

func solve1(s string, t byte) int { // t: target
    res, count := 0, 0
    for i := 0; i < len(s); i++ {
        if s[i] == t {
            count++
            res = max(res, count)
            continue
        }
        count = 0
    }
    return res
}

func solve2(s string, t1, t2 byte) int {
    res, c1, c2 := 0, 0, 0
    prev := map[int]int{}
    for i := 0; i < len(s); i++ {
        if s[i] != t1 && s[i] != t2 {
            prev = map[int]int{}
            c1, c2 = 0, 0
            continue
        }

        if s[i] == t1 {
            c1++
        } else {
            c2++
        }

        if c1 == c2 {
            l := c1 + c2 // length
            res = max(res, l)
        } else {
            diff := c1 - c2
            if p, ok := prev[diff]; ok {
                l := i - p
                res = max(res, l)
            } else {
                prev[diff] = i
            }
        }
    }
    return res
}

type Pair struct {
    x, y int
}

func solve3(s string) int {
    res, ca, cb, cc := 0, 0, 0, 0
    prev := map[Pair]int{}
    for i := 0; i < len(s); i++ {
        if s[i] == 'a' {
            ca++
        } else if s[i] == 'b' {
            cb++
        } else {
            cc++
        }

        if ca == cb && cb == cc {
            res = i + 1
        } else {
            diff := Pair{ca-cb, cb-cc}
            if p, ok := prev[diff]; ok {
                l := i - p
                res = max(res, l)
            } else {
                prev[diff] = i
            }
        }
    }
    return res
}

File: 3714-longest-balanced-substring-ii/3714-longest-balanced-substring-ii.py
class Solution:
    def longestBalanced(self, s: str) -> int:
        res = [
            solve1(s, 'a'),
            solve1(s, 'b'),
            solve1(s, 'c'),
            solve2(s, 'a', 'b'),
            solve2(s, 'a', 'c'),
            solve2(s, 'b', 'c'),
            solve3(s)
        ]

        ans = 0
        for r in res:
            ans = max(ans, r)
        return ans

# hash map, prefix sum, string
# ref: tdkkdt
# time: O(n)
# space: O(n)
    
def solve1(s, t):
    res, count = 0, 0
    for i in range(0, len(s)):
        if s[i] == t:
            count += 1
            res = max(res, count)
            continue
        count = 0
    return res

def solve2(s, t1, t2):
    res, c1, c2 = 0, 0, 0
    freq = {} # diff -> freq
    for i in range(0, len(s)):
        if s[i] != t1 and s[i] != t2:
            c1, c2 = 0, 0
            freq.clear()
            continue

        if s[i] == t1:
            c1 += 1
        else:
            c2 += 1
        
        if c1 == c2:
            res = max(res, c1 + c2)
        else:
            diff = c1 - c2
            if diff in freq:
                l = i - freq[diff]
                res = max(res, l)
            else:
                freq[diff] = i
    return res
        
def solve3(s):
    res, ca, cb, cc = 0, 0, 0, 0
    freq = {}
    for i in range(0, len(s)):
        match s[i]:
            case 'a': ca += 1
            case 'b': cb += 1
            case 'c': cc += 1
        
        if ca == cb == cc:
            res = i + 1
        else:
            diff = (ca - cb, cb - cc)
            if diff in freq:
                l = i - freq[diff]
                res = max(res, l)
            else:
                freq[diff] = i
    return res